### PR TITLE
feat(configure-async): verify configure request

### PR DESCRIPTION
**Pull request by Naccomy**
### Summary

- Introduce functionality to extract a single fact from a biscuit.
- Enable authorization of the request biscuit sent to the agent during configuration.
- Integrate a ResponseFact to attenuate the biscuit.

These updates to biscuit are designed to avoid modifying the existing codebase, ensuring no disruptions to current functionality. Subsequent PRs will address any required code updates.

### CODE REVIEW

File: theoriq/biscuit/theoriq_biscuit.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        self.request_id = request_id
+        self._body_hash = body_hash
+        self.to_addr = to_addr
+
+    @classmethod
+    def biscuit_rule(cls) -> Rule:
+        return Rule("data($req_id, $body_hash, $target_addr) <- theoriq:response($req_id, $body_hash, $target_addr)")
+
+    @classmethod
+    def from_fact(cls, fact: Fact) -> "TheoriqFact":
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:100]:
```suggestion
    def from_fact(cls, fact: Fact) -> TheoriqFact:
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        self.to_addr = verify_address(to_addr)
+
+    @classmethod
+    def biscuit_rule(cls) -> Rule:
+        return Rule(
+            "data($req_id, $body_hash, $from_addr, $target_addr) <- theoriq:request($req_id, $body_hash, $from_addr, $target_addr)"
+        )
+
+    @classmethod
+    def from_fact(cls, fact: Fact) -> "TheoriqFact":
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:66]:
```suggestion
    def from_fact(cls, fact: Fact) -> TheoriqFact:
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    @classmethod
+    @abc.abstractmethod
+    def from_fact(cls, fact: Fact) -> "TheoriqFact":
+        pass
+
+    @abc.abstractmethod
+    def to_block_builder(self) -> BlockBuilder:
+        """Convert facts to a biscuit block"""
+        pass
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:46]:
```suggestion
        """Convert facts to a biscuit block"""
    def to_block_builder(self) -> BlockBuilder:
        fact = self.get_fact()
        block_builder = BlockBuilder("")
        block_builder.add_fact(fact)
        return block_builder
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    @abc.abstractmethod
+    def biscuit_rule(cls) -> Rule:
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_fact(cls, fact: Fact) -> "TheoriqFact":
+        pass
+
+    @abc.abstractmethod
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:43]:
```suggestion
    @abc.abstractmethod
    def get_fact(self) -> Fact:
        pass
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                "req_id": str(self.request_id),
+                "body_hash": str(self.body_hash),
+                "from_addr": self.from_addr,
+                "to_addr": self.to_addr,
+            },
+        )
+
+        block_builder = BlockBuilder("")
+        block_builder.add_fact(fact)
+        return block_builder
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:83]:
```suggestion
    def get_fact(self) -> Fact:
        return Fact(
            "theoriq:request({req_id}, {body_hash}, {from_addr}, {to_addr})",
            {
                "req_id": str(self.request_id),
                "body_hash": str(self.body_hash),
                "from_addr": self.from_addr,
                "to_addr": self.to_addr,
            },
        )
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    def to_block_builder(self) -> BlockBuilder:
+        fact = Fact(
+            "theoriq:response({req_id}, {body_hash}, {to_addr})",
+            {"req_id": str(self.request_id), "body_hash": str(self._body_hash), "to_addr": self.to_addr},
+        )
+
+        block_builder = BlockBuilder("")
+        block_builder.add_fact(fact)
+        return block_builder
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:112]:
```suggestion
    def get_fact(self) -> Fact:
        return Fact(
            "theoriq:response({req_id}, {body_hash}, {to_addr})",
            {"req_id": str(self.request_id), "body_hash": str(self._body_hash), "to_addr": self.to_addr},
        )
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+
+class TheoriqBiscuit:
+    """Base class for biscuits used in Theoriq protocol"""
+
+    def __init__(self, biscuit: Biscuit) -> None:
+        self.biscuit = biscuit
+
+    @classmethod
+    def from_token(cls, token: str, public_key: str) -> "TheoriqBiscuit":
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:122]:
```suggestion
    def from_token(cls, token: str, public_key: str) -> TheoriqBiscuit:
```

File: theoriq/biscuit/theoriq_biscuit.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        public_key = public_key.removeprefix("0x")
+        biscuit = from_base64_token(token, PublicKey.from_hex(public_key))
+        return cls(biscuit)
+
+    def to_base64(self) -> str:
+        return self.biscuit.to_base64()
+
+    def attenuate(self, agent_pk: PrivateKey, fact: TheoriqFact) -> "TheoriqBiscuit":
```
Comment (aflament) [inline theoriq/biscuit/theoriq_biscuit.py:130]:
```suggestion
    def attenuate(self, agent_pk: PrivateKey, fact: TheoriqFact) -> TheoriqBiscuit:
```

File: theoriq/extra/flask/v1alpha2/flask.py [hunk 9]
```diff
@@ -14,7 +14,9 @@
 from theoriq.api.v1alpha2 import ConfigureContext
 from theoriq.api.v1alpha2.configure import AgentConfigurator
 from theoriq.api.v1alpha2.schemas import ExecuteRequestBody
-from theoriq.biscuit import TheoriqBiscuitError
+from theoriq.biscuit import PayloadHash, TheoriqBiscuitError
+from theoriq.biscuit.theoriq_biscuit import RequestFact, ResponseFact, TheoriqBiscuit
```
Comment (aflament) [inline theoriq/extra/flask/v1alpha2/flask.py:18]: add `RequestFact, ResponseFact, TheoriqBiscuit` to the biscuit `__init__.py`

**Review by aflament — CHANGES_REQUESTED** [review]

**Review by aflament — APPROVED** [review]

**Review by bvrooman — APPROVED** [review]

**Pull request merged**